--- openjdk.orig/jdk/src/share/native/sun/font/freetypeScaler.c	2010-10-15 15:19:35.000000000 +0200
+++ openjdk/jdk/src/share/native/sun/font/freetypeScaler.c	2010-10-15 15:19:35.000000000 +0200
@@ -979,7 +979,7 @@
 }
 
 
-#define FloatToF26Dot6(x) ((unsigned int) ((x)*64))
+#define FloatToF26Dot6(x) ((int) ((x)*64))
 
 static FT_Outline* getFTOutline(JNIEnv* env, jobject font2D,
         FTScalerContext *context, FTScalerInfo* scalerInfo,
