diff -u -r /jck/apache-xml/icedtea6/openjdk/jdk/src/share/classes/com/sun/org/apache/xml/internal/security/c14n/implementations/NameSpaceSymbTable.java security/c14n/implementations/NameSpaceSymbTable.java
--- openjdk/jdk/src/share/classes/com/sun/org/apache/xml/internal/security/c14n/implementations/NameSpaceSymbTable.java	2011-11-14 23:11:46.000000000 +0100
+++ openjdk/jdk/src/share/classes/com/sun/org/apache/xml/internal/security/c14n/implementations/NameSpaceSymbTable.java	2012-03-16 17:52:37.000000000 +0100
@@ -370,6 +370,7 @@
         if (cur == null || (cur.equals( obj))) {
                         return index;
         }
+        length=length-1;
         do {
                         index=index==length? 0:++index;
                         cur = set[index];
