--- openjdk.orig/hotspot/src/share/vm/runtime/arguments.cpp    2010-11-08 23:50:59.000000000 -0500
+++ openjdk/hotspot/src/share/vm/runtime/arguments.cpp        2010-11-08 23:52:53.000000000 -0500
@@ -1301,7 +1301,8 @@ void Arguments::set_ergonomics_flags() {
   if (MaxHeapSize <= max_heap_for_compressed_oops()) {
 #ifndef COMPILER1
     if (FLAG_IS_DEFAULT(UseCompressedOops) && !UseG1GC) {
-      FLAG_SET_ERGO(bool, UseCompressedOops, true);
+      // Temporarily fix rhbz#647737 -- disable compressed oops by default
+      //FLAG_SET_ERGO(bool, UseCompressedOops, true);
     }
 #endif
 #ifdef _WIN64
