diff -Nru openjdk.orig/jdk/make/common/shared/Defs-versions.gmk openjdk/jdk/make/common/shared/Defs-versions.gmk
--- openjdk.orig/jdk/make/common/shared/Defs-versions.gmk	2013-09-02 15:32:33.490472903 +0100
+++ openjdk/jdk/make/common/shared/Defs-versions.gmk	2013-09-02 15:33:00.154887749 +0100
@@ -176,7 +176,7 @@
 # Generic
 REQUIRED_ANT_VER          = 1.6.3
 REQUIRED_BOOT_VER         = 1.5
-REQUIRED_FREETYPE_VERSION = 2.3.0
+REQUIRED_FREETYPE_VERSION = 2.2.1
 REQUIRED_MAKE_VER         = 3.78
 REQUIRED_UNZIP_VER        = 5.12
 REQUIRED_ZIP_VER          = 2.2
