--- openjdk/jdk/test/java/awt/PrintJob/MultipleEnd/MultipleEnd.java.orig	2009-03-31 13:29:02.000000000 -0400
+++ openjdk/jdk/test/java/awt/PrintJob/MultipleEnd/MultipleEnd.java	2009-03-31 13:39:50.000000000 -0400
@@ -30,13 +30,17 @@
  */
 
 import java.awt.*;
+import java.awt.JobAttributes.DialogType;
 
 public class MultipleEnd {
     public static void main(String[] args) {
         new MultipleEnd().start();
     }
     public void start() {
-        new MultipleEndFrame();
+        Frame f = new MultipleEndFrame();
+        
+        // cleanup
+        f.dispose();
     }
 }
 
@@ -44,10 +48,14 @@
     public MultipleEndFrame() {
         super("MultipleEnd");
         setVisible(true);
-        PrintJob pj  = getToolkit().getPrintJob(this, "MuiltipleEnd", null);
+        JobAttributes printJobAttributes = new JobAttributes();
+        printJobAttributes.setDialog(DialogType.NONE);
+        PrintJob pj  = getToolkit().getPrintJob(this, "MuiltipleEnd", printJobAttributes, null);
         if (pj != null) {
             pj.end();
             pj.end();
+        } else {
+            throw new RuntimeException("getPrintJob returned null");
         }
     }
 }
