diff -r 8e41d25d7665 openjdk/control/make/make/hotspot-rules.gmk
--- openjdk/control/make/make/hotspot-rules.gmk	Fri Jun 20 13:26:10 2008 +0100
+++ openjdk/make/hotspot-rules.gmk	Fri Jun 20 13:26:20 2008 +0100
@@ -73,7 +73,11 @@
 endif
 
 ifeq ($(ZERO_BUILD), true)
-  HOTSPOT_TARGET := $(HOTSPOT_TARGET)zero
+  ifeq ($(SHARK_BUILD), true)
+    HOTSPOT_TARGET := $(HOTSPOT_TARGET)shark
+  else
+    HOTSPOT_TARGET := $(HOTSPOT_TARGET)zero
+  endif
 endif
 
 HOTSPOT_BUILD_ARGUMENTS += $(COMMON_BUILD_ARGUMENTS)
