--- openjdk/jdk/test/sun/tools/jhat/HatRun.java.orig	2009-03-20 09:55:39.000000000 -0400
+++ openjdk/jdk/test/sun/tools/jhat/HatRun.java	2009-03-20 09:56:48.000000000 -0400
@@ -161,6 +161,7 @@
     public void runit(String class_name, String vm_options[])
     {
         String jre_home  = System.getProperty("java.home");
+        String classpath = System.getProperty("java.class.path");
         String sdk_home  = (jre_home.endsWith("jre") ?
                             (jre_home + File.separator + "..") :
                             jre_home );
@@ -170,27 +171,31 @@
                            os_arch.equals("amd64");
         String isa_dir   = d64?(File.separator+os_arch):"";
         String java      = jre_home
-                             + File.separator + "bin" + isa_dir
+                             + File.separator + "bin"
                              + File.separator + "java";
         String jhat      = sdk_home + File.separator + "bin"
                            + File.separator + "jhat";
         /* Array of strings to be passed in for exec:
          *   1. java
-         *   2. -Dtest.classes=.
-         *   3. -d64                 (optional)
-         *   4. -Xcheck:jni          (Just because it finds bugs)
-         *   5. -Xverify:all         (Make sure verification is on full blast)
-         *   6. -agent
+         *   2. -cp 
+         *   3. classpath
+         *   4. -Dtest.classes=.
+         *   5. -d64                 (optional)
+         *   6. -Xcheck:jni          (Just because it finds bugs)
+         *   7. -Xverify:all         (Make sure verification is on full blast)
+         *   8. -agent
          *       vm_options
-         *   7+i. classname
+         *   9+i. classname
          */
         int nvm_options = 0;
         if ( vm_options != null ) nvm_options = vm_options.length;
-        String cmd[]     = new String[1 + (d64?1:0) + 5 + nvm_options];
+        String cmd[]     = new String[1 + (d64?1:0) + 7 + nvm_options];
         int i,j;
 
         i = 0;
         cmd[i++] = java;
+        cmd[i++] = "-cp";
+        cmd[i++] = classpath;
         cmd[i++] = "-Dtest.classes=" + cdir;
         if ( d64 ) {
             cmd[i++] = "-d64";
