diff -r 348fce38de3f src/share/classes/java/util/SimpleTimeZone.java
--- openjdk/jdk/src/share/classes/java/util/SimpleTimeZone.java	Fri Jun 26 19:50:44 2009 +0400
+++ openjdk/jdk/src/share/classes/java/util/SimpleTimeZone.java	Sat Aug 29 22:36:41 2009 +0200
@@ -1372,7 +1372,7 @@
                 throw new IllegalArgumentException(
                         "Illegal start month " + startMonth);
             }
-            if (startTime < 0 || startTime >= millisPerDay) {
+            if (startTime < 0 || startTime > millisPerDay) {
                 throw new IllegalArgumentException(
                         "Illegal start time " + startTime);
             }
@@ -1419,7 +1419,7 @@
                 throw new IllegalArgumentException(
                         "Illegal end month " + endMonth);
             }
-            if (endTime < 0 || endTime >= millisPerDay) {
+            if (endTime < 0 || endTime > millisPerDay) {
                 throw new IllegalArgumentException(
                         "Illegal end time " + endTime);
             }
