--- oldMessageUtils.c	2009-06-25 16:39:27.000000000 -0400
+++ openjdk/jdk/src/share/native/sun/misc/MessageUtils.c	2009-06-25 16:40:14.000000000 -0400
@@ -39,6 +39,11 @@
     int i;
     const jchar *sAsArray;
 
+    if (s == NULL) {
+      s = (*env)->NewStringUTF(env, "null");
+      if (s == NULL) return;
+    }
+
     sAsArray = (*env)->GetStringChars(env, s, NULL);
     length = (*env)->GetStringLength(env, s);
     sConverted = (char *) malloc(length + 1);
