--- old/src/share/classes/sun/net/httpserver/Request.java	Wed Mar  4 03:29:39 2009
+++ openjdk/jdk/src/share/classes/sun/net/httpserver/Request.java	Wed Mar  4 03:29:39 2009
@@ -52,6 +52,9 @@
         os = rawout;
         do {
             startLine = readLine();
+            if (startLine == null) {
+                return;
+            }
             /* skip blank lines */
         } while (startLine.equals (""));
     }
--- old/src/share/classes/sun/net/httpserver/ServerImpl.java	Wed Mar  4 03:29:41 2009
+++ openjdk/jdk/src/share/classes/sun/net/httpserver/ServerImpl.java	Wed Mar  4 03:29:41 2009
@@ -437,6 +437,7 @@
                         rawin = sslStreams.getInputStream();
                         rawout = sslStreams.getOutputStream();
                         engine = sslStreams.getSSLEngine();
+                        connection.sslStreams = sslStreams;
                     } else {
                         rawin = new BufferedInputStream(
                             new Request.ReadStream (
@@ -446,6 +447,8 @@
                             ServerImpl.this, chan
                         );
                     }
+                    connection.raw = rawin;
+                    connection.rawout = rawout;
                 }
                 Request req = new Request (rawin, rawout);
                 requestLine = req.requestLine();
