--- openjdk/jdk/test/java/awt/Graphics/DrawImageBG/SystemBgColorTest.java.orig	2009-03-30 14:21:10.000000000 -0400
+++ openjdk/jdk/test/java/awt/Graphics/DrawImageBG/SystemBgColorTest.java	2009-03-30 14:21:18.000000000 -0400
@@ -59,7 +59,6 @@
         if (errcount > 0) {
             throw new RuntimeException(errcount+" errors");
         }
-        System.exit(0); // For 1.3 and earlier VMs...
     }
 
     static int cmap[] = {
