--- openjdk/jdk/test/java/awt/PrintJob/EdgeTest/EdgeTest.java.orig	2009-03-31 16:49:27.000000000 -0400
+++ openjdk/jdk/test/java/awt/PrintJob/EdgeTest/EdgeTest.java	2009-03-31 16:49:54.000000000 -0400
@@ -26,6 +26,7 @@
  * @bug 4092755
  * @summary Verifies that (0, 0) is the upper-left corner of the page, not
  *          the upper-left corner adjusted for the margins.
+ * @run main/manual=yesno EdgeTest
  * @author dpm
  */
 
