--- /dev/null	2009-03-12 10:05:36.797002285 -0400
+++ openjdk/jdk/test/com/sun/org/apache/xml/internal/serializer/XMLStackOverflowBug.java	2009-03-13 16:10:05.000000000 -0400
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2009 Red Hat, Inc.  All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
+ * CA 95054 USA or visit www.sun.com if you need additional information or
+ * have any questions.
+ */
+
+/*
+ * @test
+ * @summary Check that the xml encoder doesnt cause a StackOverflowError
+ *
+ */
+
+import java.io.IOException;
+
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.sax.SAXTransformerFactory;
+import javax.xml.transform.sax.TransformerHandler;
+import javax.xml.transform.stream.StreamResult;
+
+import org.xml.sax.SAXException;
+
+public class XMLStackOverflowBug {
+
+    public static void main(String[] args)
+            throws TransformerConfigurationException, IOException, SAXException {
+
+        SAXTransformerFactory stf = (SAXTransformerFactory) TransformerFactory
+                .newInstance();
+        TransformerHandler ser = stf.newTransformerHandler();
+        ser.setResult(new StreamResult(System.out));
+
+        StringBuilder sb = new StringBuilder(4096); 
+        for (int x = 4096; x > 0; x--) {
+            sb.append((char)x);
+        }
+        ser.characters(sb.toString().toCharArray(), 0, sb.toString().toCharArray().length);
+        ser.endDocument();
+    }
+}
+
diff -Nru openjdk.orig/jaxp/build.properties openjdk/jaxp/build.properties
--- openjdk.orig/jaxp/build.properties	2009-12-08 17:42:33.000000000 +0000
+++ openjdk/jaxp/build.properties	2009-12-08 17:43:03.000000000 +0000
@@ -73,6 +73,9 @@
 # Where patches to drop bundle sources live
 patches.dir=patches
 
+# Patches to apply
+jaxp_src.patch.list=xml-encodinginfo.patch
+
 # Sanity information
 sanity.info= Sanity Settings:${line.separator}\
   ant.home=${ant.home}${line.separator}\
diff -Nru openjdk.orig/jaxp/patches/jaxp_src/xml-encodinginfo.patch openjdk/jaxp/patches/jaxp_src/xml-encodinginfo.patch
--- openjdk.orig/jaxp/patches/jaxp_src/xml-encodinginfo.patch	1970-01-01 01:00:00.000000000 +0100
+++ openjdk/jaxp/patches/jaxp_src/xml-encodinginfo.patch	2009-12-08 17:41:58.000000000 +0000
@@ -0,0 +1,18 @@
+diff -Nru src/com/sun/org/apache/xml/internal/serializer/EncodingInfo.java src.new/com/sun/org/apache/xml/internal/serializer/EncodingInfo.java
+--- src/com/sun/org/apache/xml/internal/serializer/EncodingInfo.java	2009-10-27 21:54:16.000000000 +0000
++++ src.new/com/sun/org/apache/xml/internal/serializer/EncodingInfo.java	2009-12-08 17:40:14.000000000 +0000
+@@ -326,9 +326,11 @@
+             m_last = last;
+ 
+             // Set the range of unicode values that this object
+-            // explicitly manages
+-            m_explFirst = codePoint;
+-            m_explLast = codePoint + (RANGE-1);
++            // explicitly manages. Align the explicitly managed values
++            // to RANGE so multiple EncodingImpl objects dont manage the same 
++            // values.
++            m_explFirst = codePoint / RANGE * RANGE;
++            m_explLast = m_explFirst + (RANGE-1);
+ 
+             m_encoding = encoding;
+ 
