--- openjdk6.orig/jdk/src/share/classes/sun/util/calendar/ZoneInfoFile.java	Sat Mar 15 13:43:05 2008 -0400
+++ openjdk/jdk/src/share/classes/sun/util/calendar/ZoneInfoFile.java	Fri Aug 21 11:34:56 2009 +0200
@@ -465,6 +465,27 @@
      */
     public static final byte    TAG_TZDataVersion = 68;
 
+    // Cached location of the TZDATA files
+    private static final String JAVAZI_DIR = setup_JAVAZI_DIR();
+
+    private static String setup_JAVAZI_DIR() {
+      try {
+        final String dir = AccessController.doPrivileged
+          (new sun.security.action.GetPropertyAction("user.zoneinfo.dir"));
+            return AccessController.doPrivileged
+              (new PrivilegedExceptionAction<String>() {
+                public String run() {
+                  File f = new File(dir, "ZoneInfoMappings");
+                  if (f.exists())
+                    return dir;
+                  return null;
+                }
+              });
+        } catch (PrivilegedActionException e) {
+          return null;
+        }
+    }
+
     /**
      * Excluded zones item tag. (Added in Mustang)
      */
@@ -1021,11 +1042,17 @@
         byte[] buffer = null;
 
         try {
-            String homeDir = AccessController.doPrivileged(
-                new sun.security.action.GetPropertyAction("java.home"));
-            final String fname = homeDir + File.separator + "lib" + File.separator
-                                 + "zi" + File.separator + fileName;
-            buffer = (byte[]) AccessController.doPrivileged(new PrivilegedExceptionAction() {
+
+	  String zi_dir = JAVAZI_DIR;
+	  if (zi_dir == null) {
+	    // Fall back to JDK-supplied tzdata
+	    String homeDir = (String) AccessController.doPrivileged(new sun.security.action.GetPropertyAction("java.home"));
+	    zi_dir = homeDir + File.separator + "lib" + File.separator
+	      + "zi";
+	  }
+
+          final String fname =  zi_dir + File.separator + fileName;
+	  buffer = (byte[]) AccessController.doPrivileged(new PrivilegedExceptionAction() {
                 public Object run() throws IOException {
                     File file = new File(fname);
                     if (!file.canRead()) {
--- oldopenjdk6/hotspot/src/os/linux/vm/os_linux.cpp	2008-03-26 05:07:22.000000000 -0400
+++ openjdk/hotspot/src/os/linux/vm/os_linux.cpp	2008-05-08 11:03:17.000000000 -0400
@@ -376,6 +381,11 @@
     }
   }
 
+  SystemProperty* sp = Arguments::system_properties();
+  // Use the system zoneinfo files, if present
+  Arguments::PropertyList_add (&sp,
+                           "user.zoneinfo.dir", "/usr/share/javazi");
+
 #undef malloc
 #undef getenv
 #undef EXTENSIONS_DIR
