--- openjdk7/jdk/src/share/classes/java/util/TimeZone.java	2009-08-04 18:54:13.026104895 +0200
+++ openjdk/jdk/src/share/classes/java/util/TimeZone.java	2009-08-04 18:53:09.035985858 +0200
@@ -602,11 +602,7 @@
                 }
             });
 
-        if (hasPermission()) {
-            defaultTimeZone = tz;
-        } else {
-            defaultZoneTL.set(tz);
-        }
+        defaultTimeZone = tz;
         return tz;
     }
 
@@ -637,6 +633,7 @@
         if (hasPermission()) {
             synchronized (TimeZone.class) {
                 defaultTimeZone = zone;
+                defaultZoneTL.set(null);
             }
         } else {
             defaultZoneTL.set(zone);
