--- /dev/null	1970-01-01 01:00:00.000000000 +0100
+++ openjdk/jdk/test/java/awt/Insets/WindowWithWarningTest/applet.policy	2010-11-26 18:01:32.000000000 +0100
@@ -0,0 +1,13 @@
+
+// Standard extensions get all permissions by default
+
+grant codeBase "file:${{java.ext.dirs}}/*" {
+	permission java.security.AllPermission;
+};
+
+// default permissions granted to all domains
+
+grant { 
+	permission java.awt.AWTPermission "createRobot", "enabled";
+};
+
--- openjdk-old/jdk/test/java/awt/Insets/WindowWithWarningTest/WindowWithWarningTest.html	2010-06-21 23:15:49.000000000 +0200
+++ openjdk/jdk/test/java/awt/Insets/WindowWithWarningTest/WindowWithWarningTest.html	2010-11-26 18:29:25.622555000 +0100
@@ -4,7 +4,7 @@
         @bug 6391770
         @summary Content of the Window should be laid out in the area left after WarningWindow was added.
         @author Yuri Nesterenko
-        @run applet WindowWithWarningTest.html
+        @run applet/othervm/policy=applet.policy WindowWithWarningTest.html
 -->
 
   <head>
--- openjdk-old/jdk/test/java/awt/Insets/WindowWithWarningTest/WindowWithWarningTest.java	2010-06-21 23:15:49.000000000 +0200
+++ openjdk/jdk/test/java/awt/Insets/WindowWithWarningTest/WindowWithWarningTest.java	2010-11-26 18:10:15.000000000 +0100
@@ -53,6 +53,7 @@
 import java.awt.*;
 import java.awt.event.*;
 import javax.swing.*;
+import java.security.*;
 
 //Automated tests should run as applet tests if possible because they
 // get their environments cleaned up, including AWT threads, any
@@ -91,12 +92,6 @@
     public void start ()
     {
         //Get things going.  Request focus, set size, et cetera
-        System.setSecurityManager( new SecurityManager() {
-        // deny AWTPermission("showWindowWithoutWarningBanner")
-            public boolean checkTopLevelWindow(Object window) {
-                return false;
-            }
-         });
         JFrame frame = new JFrame("Window Test");
         frame.setBounds(50, 50, 200, 200);
         frame.show();
