--- openjdk-old/jdk/test/java/awt/FontClass/CreateFont/DeleteFont.sh	2010-06-21 23:15:49.000000000 +0200
+++ openjdk/jdk/test/java/awt/FontClass/CreateFont/DeleteFont.sh	2010-11-15 14:13:13.000000000 +0100
@@ -46,7 +46,7 @@
 
 cd ${TESTCLASSES}
 
-numfiles0=`ls ${TESTCLASSES} | wc -l`
+numfiles0=`ls ${TESTCLASSES} | grep -v "^hsperfdata*" | wc -l`
 ${TESTJAVA}/bin/java -Djava.io.tmpdir=${TESTCLASSES} DeleteFont
 
 if [ $? -ne 0 ]
@@ -55,7 +55,7 @@
       exit 1
 fi
 
-numfiles1=`ls ${TESTCLASSES} | wc -l`
+numfiles1=`ls ${TESTCLASSES} | grep -v "^hsperfdata*" | wc -l`
 
 if [ $numfiles0 -ne $numfiles1 ]
     then
