--- openjdk.orig/jdk/test/sun/net/www/protocol/file/EncodedMultiByteChar.java	2011-02-28 17:07:06.000000000 +0100
+++ openjdk/jdk/test/sun/net/www/protocol/file/EncodedMultiByteChar.java	2011-06-14 13:21:36.000000000 +0200
@@ -61,6 +61,9 @@
         } catch (IOException e) {
             System.out.println("Failed to create test file ");
             e.printStackTrace();
+            if (file != null) {
+                file.delete();
+            }
             return;
         }
 
@@ -81,5 +84,11 @@
         } catch (IOException ioe) {
             ioe.printStackTrace();
         }
+        finally {
+            if (file != null) {
+                file.delete();
+            }
+        }
+        System.out.println("test OK");
     }
 }
