--- openjdk.orig/jdk/test/java/util/zip/ConstructDeflaterInput_original.java	2011-07-05 20:36:25.000000000 +0200
+++ openjdk/jdk/test/java/util/zip/ConstructDeflaterInput.java	2011-07-11 15:35:00.000000000 +0200
@@ -25,6 +25,7 @@
  * @test
  * @bug 4679743
  * @summary Test parts of DeflaterInputStream code that don't really do I/O.
+ * @run main/othervm ConstructDeflaterInput
  */
 
 import java.io.*;
