--- openjdk.orig/hotspot/test/runtime/7020373/Test7020373.sh	2011-07-08 15:04:32.000000000 +0200
+++ openjdk/hotspot/test/runtime/7020373/Test7020373.sh	2011-07-08 15:01:13.000000000 +0200
@@ -3,6 +3,7 @@
 ##
 ## @test
 ## @bug 7020373
+## @ignore The original test contains binary jar-file
 ## @key cte_test
 ## @summary JSR rewriting can overflow memory address size variables
 ## @run shell Test7020373.sh
