--- openjdk/jdk/src/share/classes/sun/awt/SunToolkit.orig.java	2011-03-11 10:48:29.000000000 +0100
+++ openjdk/jdk/src/share/classes/sun/awt/SunToolkit.java	2011-03-11 11:24:33.000000000 +0100
@@ -65,8 +65,7 @@

     /* Load debug settings for native code */
     static {
-        String nativeDebug = System.getProperty("sun.awt.nativedebug");
-        if ("true".equalsIgnoreCase(nativeDebug)) {
+        if (AccessController.doPrivileged(new GetBooleanAction("sun.awt.nativedebug"))) {
             DebugSettings.init();
         }
     };
