--- openjdk.orig/jdk/make/javax/swing/Makefile	2010-06-13 20:19:57.324427433 +0100
+++ openjdk/jdk/make/javax/swing/Makefile	2010-06-13 20:19:57.161043101 +0100
@@ -35,6 +35,8 @@
 NIMBUS_GENERATOR_JAR = $(BUILDTOOLJARDIR)/generatenimbus.jar
 
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 #
 # Files
--- openjdk.orig/jdk/make/javax/swing/plaf/Makefile	2010-06-13 20:19:58.277416821 +0100
+++ openjdk/jdk/make/javax/swing/plaf/Makefile	2010-06-13 20:19:58.121793675 +0100
@@ -29,6 +29,8 @@
 SWING_SRC = $(SHARE_SRC)/classes/javax/swing
 
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 #
 # Files
