--- openjdk-old/jdk/make/sun/xawt/mapfile-vers	2010-08-16 16:54:21.000000000 +0200
+++ openjdk/jdk/make/sun/xawt/mapfile-vers	2010-08-16 16:54:21.000000000 +0200
@@ -300,6 +300,7 @@
         Java_java_awt_FileDialog_initIDs;
         Java_sun_awt_X11_XWindow_initIDs;
 
+        Java_sun_java2d_opengl_OGLContext_getOGLIdString;
         Java_sun_java2d_opengl_OGLMaskFill_maskFill;
         Java_sun_java2d_opengl_OGLRenderer_drawPoly;
         Java_sun_java2d_opengl_OGLRenderQueue_flushBuffer;
