--- openjdk.orig/hotspot/src/share/vm/ci/ciTypeFlow.cpp	2011-01-25 22:57:24.000000000 +0000
+++ openjdk/hotspot/src/share/vm/ci/ciTypeFlow.cpp	2011-02-24 00:28:58.000000000 +0000
@@ -37,6 +37,9 @@
 #include "memory/allocation.inline.hpp"
 #include "runtime/deoptimization.hpp"
 #include "utilities/growableArray.hpp"
+#ifdef SHARK
+#include "shark/shark_globals.hpp"
+#endif
 
 // ciTypeFlow::JsrSet
 //
--- openjdk.orig/hotspot/make/linux/makefiles/vm.make	2011-02-23 21:28:45.000000000 +0000
+++ openjdk/hotspot/make/linux/makefiles/vm.make	2011-02-24 02:17:30.000000000 +0000
@@ -139,13 +139,15 @@
 COMPILER2_PATHS += $(GAMMADIR)/src/share/vm/libadt
 COMPILER2_PATHS +=  $(GENERATED)/adfiles
 
+SHARK_PATHS := $(GAMMADIR)/src/share/vm/shark
+
 # Include dirs per type.
 Src_Dirs/CORE      := $(CORE_PATHS)
 Src_Dirs/COMPILER1 := $(CORE_PATHS) $(COMPILER1_PATHS)
 Src_Dirs/COMPILER2 := $(CORE_PATHS) $(COMPILER2_PATHS)
 Src_Dirs/TIERED    := $(CORE_PATHS) $(COMPILER1_PATHS) $(COMPILER2_PATHS)
 Src_Dirs/ZERO      := $(CORE_PATHS)
-Src_Dirs/SHARK     := $(CORE_PATHS)
+Src_Dirs/SHARK     := $(CORE_PATHS) $(SHARK_PATHS)
 Src_Dirs := $(Src_Dirs/$(TYPE))
 
 COMPILER2_SPECIFIC_FILES := opto libadt bcEscapeAnalyzer.cpp chaitin\* c2_\* runtime_\*
--- openjdk.orig/hotspot/make/linux/makefiles/launcher.make	2011-02-22 16:51:24.000000000 +0000
+++ openjdk/hotspot/make/linux/makefiles/launcher.make	2011-02-24 03:01:36.000000000 +0000
@@ -52,6 +52,10 @@
   LAUNCHER.o                 = launcher.o
   LFLAGS_LAUNCHER           += -L `pwd`
   LIBS_LAUNCHER             += -l$(JVM) $(LIBS)
+  ifeq ($(SHARK_BUILD), true)
+    LFLAGS_LAUNCHER += $(LLVM_LDFLAGS)
+    LIBS_LAUNCHER   += $(LLVM_LIBS)
+  endif
 endif
 
 LINK_LAUNCHER = $(LINK.c)
