--- openjdk.orig/hotspot/src/share/vm/ci/ciTypeFlow.cpp	2011-01-25 22:57:24.000000000 +0000
+++ openjdk/hotspot/src/share/vm/ci/ciTypeFlow.cpp	2011-02-24 00:28:58.000000000 +0000
@@ -37,6 +37,9 @@
 #include "memory/allocation.inline.hpp"
 #include "runtime/deoptimization.hpp"
 #include "utilities/growableArray.hpp"
+#ifdef SHARK
+#include "shark/shark_globals.hpp"
+#endif
 
 // ciTypeFlow::JsrSet
 //
