Index: openjdk/hotspot/src/share/vm/interpreter/bytecodeInterpreter.cpp
===================================================================
--- openjdk.orig/hotspot/src/share/vm/interpreter/bytecodeInterpreter.cpp Sun Apr 03 12:00:54 2011 +0200
+++ openjdk/hotspot/src/share/vm/interpreter/bytecodeInterpreter.cpp Mon Apr 04 03:02:00 2011 -0700
@@ -568,7 +568,7 @@ BytecodeInterpreter::run(interpreterStat
 /* 0xDC */ &&opc_default, &&opc_default, &&opc_default, &&opc_default,
  
 /* 0xE0 */ &&opc_default, &&opc_default, &&opc_default, &&opc_default,
-/* 0xE4 */ &&opc_default, &&opc_return_register_finalizer, &&opc_default, &&opc_default,
+/* 0xE4 */ &&opc_default, &&opc_default, &&opc_default, &&opc_return_register_finalizer,
 /* 0xE8 */ &&opc_default, &&opc_default, &&opc_default, &&opc_default,
 /* 0xEC */ &&opc_default, &&opc_default, &&opc_default, &&opc_default,
  
# HG changeset patch
# User twisti
# Date 1300957910 25200
# Node ID 151da0c145a8e4de263740273ec10c0eb46b30d0
# Parent 9dc311b8473e0259a5de3f9358ca94de3990d692
7030207: Zero tweak to remove accidentally incorporated code
Summary: IcedTea contains a now-unmaintained ARM-specific interpreter and part of that interpreter was accidentally incorporated in one of the webrevs when Zero was initially imported.
Reviewed-by: twisti
Contributed-by: Gary Benson <gbenson@redhat.com>

--- openjdk.orig/hotspot/src/share/vm/interpreter/bytecodeInterpreter.cpp	Mon Mar 21 11:28:14 2011 -0700
+++ openjdk/hotspot/src/share/vm/interpreter/bytecodeInterpreter.cpp	Thu Mar 24 02:11:50 2011 -0700
@@ -2383,17 +2383,6 @@ run:
       }
 
       DEFAULT:
-#ifdef ZERO
-          // Some zero configurations use the C++ interpreter as a
-          // fallback interpreter and have support for platform
-          // specific fast bytecodes which aren't supported here, so
-          // redispatch to the equivalent non-fast bytecode when they
-          // are encountered.
-          if (Bytecodes::is_defined((Bytecodes::Code)opcode)) {
-              opcode = (jubyte)Bytecodes::java_code((Bytecodes::Code)opcode);
-              goto opcode_switch;
-          }
-#endif
           fatal(err_msg("Unimplemented opcode %d = %s", opcode,
                         Bytecodes::name((Bytecodes::Code)opcode)));
           goto finish;

