diff -r 55b9f498dbce make/linux/makefiles/gcc.make
--- openjdk/hotspot/make/linux/makefiles/gcc.make	Thu Feb 10 16:24:29 2011 -0800
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	Fri Feb 11 18:09:21 2011 -0500
@@ -196,7 +196,7 @@
 SHARED_FLAG = -shared
 
 # Keep symbols even they are not used
-AOUT_FLAGS += -export-dynamic
+AOUT_FLAGS += -Wl,-export-dynamic
 
 #------------------------------------------------------------------------
 # Debug flags
