diff -r d835d67bfb88 src/share/vm/shark/sharkCompiler.cpp
--- openjdk/hotspot/src/share/vm/shark/sharkCompiler.cpp	Thu Dec 05 20:12:11 2013 -0500
+++ openjdk/hotspot/src/share/vm/shark/sharkCompiler.cpp	Thu Dec 26 14:22:34 2013 +0000
@@ -212,7 +212,6 @@
                        &inc_table,
                        this,
                        env->comp_level(),
-                       false,
                        false);
 }
 
