diff -Nru openjdk.orig/hotspot/src/cpu/zero/vm/disassembler_zero.hpp openjdk/hotspot/src/cpu/zero/vm/disassembler_zero.hpp
--- openjdk.orig/hotspot/src/cpu/zero/vm/disassembler_zero.hpp	2010-05-28 11:10:31.000000000 +0100
+++ openjdk/hotspot/src/cpu/zero/vm/disassembler_zero.hpp	2010-06-17 19:29:24.000000000 +0100
@@ -23,13 +23,10 @@
  *
  */
 
-// The disassembler prints out zero code annotated
-// with Java specific information.
-
   static int pd_instruction_alignment() {
-    ShouldNotCallThis();
+    return 1;
   }
 
   static const char* pd_cpu_opts() {
-    ShouldNotCallThis();
+    return "";
   }
diff -Nru openjdk.orig/hotspot/src/cpu/zero/vm/globals_zero.hpp openjdk/hotspot/src/cpu/zero/vm/globals_zero.hpp
--- openjdk.orig/hotspot/src/cpu/zero/vm/globals_zero.hpp	2010-05-28 11:10:31.000000000 +0100
+++ openjdk/hotspot/src/cpu/zero/vm/globals_zero.hpp	2010-06-17 19:29:07.000000000 +0100
@@ -35,6 +35,7 @@
 define_pd_global(bool,  UncommonNullCast,     true);
 
 define_pd_global(intx,  CodeEntryAlignment,   32);
+define_pd_global(intx,  OptoLoopAlignment,    16);
 define_pd_global(intx,  InlineFrequencyCount, 100);
 define_pd_global(intx,  PreInflateSpin,       10);
 
