diff -ur a/openjdk/corba/make/common/Defs-linux.gmk b/openjdk/corba/make/common/Defs-linux.gmk
--- a/openjdk/corba/make/common/Defs-linux.gmk	2013-12-10 22:33:53.803394885 +0000
+++ b/openjdk/corba/make/common/Defs-linux.gmk	2013-12-10 22:37:11.261367112 +0000
@@ -301,7 +301,7 @@
 override LIBTHREAD               =
 override MOOT_PRIORITIES         = true
 override NO_INTERRUPTIBLE_IO     = true
-override OPENWIN_HOME            = /usr/X11R6
+override OPENWIN_HOME            = /opt/Xorg
 ifeq ($(ARCH), amd64)
 override OPENWIN_LIB             = $(OPENWIN_HOME)/lib64
 else
Only in b/openjdk/hotspot/test: serviceability
diff -ur a/openjdk/jdk/make/common/Defs-linux.gmk b/openjdk/jdk/make/common/Defs-linux.gmk
--- a/openjdk/jdk/make/common/Defs-linux.gmk	2013-12-10 22:33:55.319417838 +0000
+++ b/openjdk/jdk/make/common/Defs-linux.gmk	2013-12-10 22:37:11.279367381 +0000
@@ -331,7 +331,7 @@
 override LIBTHREAD               =
 override MOOT_PRIORITIES         = true
 override NO_INTERRUPTIBLE_IO     = true
-override OPENWIN_HOME            = /usr/X11R6
+override OPENWIN_HOME            = /opt/Xorg
 ifeq ($(ARCH), amd64)
 override OPENWIN_LIB             = $(OPENWIN_HOME)/lib64
 else
diff -ur a/openjdk/jdk/make/sun/awt/mawt.gmk b/openjdk/jdk/make/sun/awt/mawt.gmk
--- a/openjdk/jdk/make/sun/awt/mawt.gmk	2013-12-10 22:33:53.543390949 +0000
+++ b/openjdk/jdk/make/sun/awt/mawt.gmk	2013-12-10 22:37:11.292367576 +0000
@@ -146,11 +146,11 @@
             LIBXT = -lXt
         else
         # Allows for builds on Debian GNU Linux, X11 is in a different place 
-	    LIBXT = $(firstword $(wildcard /usr/X11R6/lib/libXt.a) \
+	    LIBXT = $(firstword $(wildcard /opt/Xorg/lib/libXt.a) \
                             $(wildcard /usr/lib/libXt.a))
-	    LIBSM = $(firstword $(wildcard /usr/X11R6/lib/libSM.a) \
+	    LIBSM = $(firstword $(wildcard /opt/Xorg/lib/libSM.a) \
                             $(wildcard /usr/lib/libSM.a))
-	    LIBICE = $(firstword $(wildcard /usr/X11R6/lib/libICE.a) \
+	    LIBICE = $(firstword $(wildcard /opt/Xorg/lib/libICE.a) \
                              $(wildcard /usr/lib/libICE.a))
 	endif
     endif
@@ -225,8 +225,7 @@
 
 ifeq ($(PLATFORM), linux)
 # Checking for the X11/extensions headers at the additional location
-  CPPFLAGS += -I/X11R6/include/X11/extensions \
-              -I/usr/include/X11/extensions
+  CPPFLAGS += -I=/opt/Xorg/include/X11/extensions -I=/opt/Xorg/include
 endif
 
 ifeq ($(PLATFORM), solaris)
diff -ur a/openjdk/jdk/make/sun/font/Makefile b/openjdk/jdk/make/sun/font/Makefile
--- a/openjdk/jdk/make/sun/font/Makefile	2013-12-10 22:33:55.252416823 +0000
+++ b/openjdk/jdk/make/sun/font/Makefile	2013-12-10 22:37:11.297367651 +0000
@@ -163,7 +163,7 @@
 ifndef OPENJDK
   CPPFLAGS += -I$(CLOSED_SRC)/share/native/$(PKGDIR)/t2k
 else
-  CPPFLAGS += -I$(FREETYPE_HEADERS_PATH) -I$(FREETYPE_HEADERS_PATH)/freetype2
+  CPPFLAGS += -I=$(FREETYPE_HEADERS_PATH) -I=$(FREETYPE_HEADERS_PATH)/freetype2 -I=/opt/Xorg/include
   OTHER_LDLIBS += $(FREETYPE_LIB)
 endif
 
diff -ur a/openjdk/jdk/make/sun/xawt/Makefile b/openjdk/jdk/make/sun/xawt/Makefile
--- a/openjdk/jdk/make/sun/xawt/Makefile	2013-12-10 22:33:55.379418746 +0000
+++ b/openjdk/jdk/make/sun/xawt/Makefile	2013-12-10 22:37:11.308367815 +0000
@@ -130,8 +130,7 @@
 
 ifeq ($(PLATFORM), linux)
 # Allows for builds on Debian GNU Linux, X11 is in a different place 
-  CPPFLAGS += -I/usr/X11R6/include/X11/extensions \
-              -I/usr/include/X11/extensions \
+  CPPFLAGS += -I=/opt/Xorg/include/X11/extensions \
               -I$(MOTIF_DIR)/include \
               -I$(OPENWIN_HOME)/include 
 endif
