Index: hotspot/src/share/vm/runtime/frame.cpp
===================================================================
--- openjdk/hotspot.orig/src/share/vm/runtime/frame.cpp	2010-02-17 04:11:10.000000000 +0100
+++ openjdk/hotspot/src/share/vm/runtime/frame.cpp	2010-03-25 10:06:18.334380336 +0100
@@ -1205,6 +1205,10 @@
          if (is_interpreted_frame())    { oops_interpreted_do(f, map, use_interpreter_oop_map_cache);
   } else if (is_entry_frame())          { oops_entry_do      (f, map);
   } else if (CodeCache::contains(pc())) { oops_code_blob_do  (f, map);
+#ifdef SHARK
+  } else if (is_fake_stub_frame()) {
+    // nothing to do
+#endif // SHARK
   } else {
     ShouldNotReachHere();
   }
