diff -Nru openjdk.orig/hotspot/make/linux/makefiles/rules.make openjdk/hotspot/make/linux/makefiles/rules.make
--- openjdk.orig/hotspot/make/linux/makefiles/rules.make	2013-02-11 23:08:47.000000000 +0000
+++ openjdk/hotspot/make/linux/makefiles/rules.make	2013-02-15 14:53:43.564311236 +0000
@@ -128,7 +128,7 @@
 # Settings for javac
 BOOT_SOURCE_LANGUAGE_VERSION = 6
 BOOT_TARGET_CLASS_VERSION = 6
-JAVAC_FLAGS = -g -encoding ascii
+JAVAC_FLAGS = -g -encoding ascii -J-Xmx512m
 BOOTSTRAP_JAVAC_FLAGS = $(JAVAC_FLAGS) -source $(BOOT_SOURCE_LANGUAGE_VERSION) -target $(BOOT_TARGET_CLASS_VERSION)
 
 # With parallel makes, print a message at the end of compilation.
diff -Nru openjdk.orig/hotspot/make/solaris/makefiles/rules.make openjdk/hotspot/make/solaris/makefiles/rules.make
--- openjdk.orig/hotspot/make/solaris/makefiles/rules.make	2013-02-11 23:08:47.000000000 +0000
+++ openjdk/hotspot/make/solaris/makefiles/rules.make	2013-02-15 14:54:13.620795503 +0000
@@ -120,7 +120,7 @@
 # Settings for javac
 BOOT_SOURCE_LANGUAGE_VERSION = 6
 BOOT_TARGET_CLASS_VERSION = 6
-JAVAC_FLAGS = -g -encoding ascii
+JAVAC_FLAGS = -g -encoding ascii -J-Xmx512m
 BOOTSTRAP_JAVAC_FLAGS = $(JAVAC_FLAGS) -source $(BOOT_SOURCE_LANGUAGE_VERSION) -target $(BOOT_TARGET_CLASS_VERSION)
 
 # With parallel makes, print a message at the end of compilation.
diff -Nru openjdk.orig/hotspot/make/windows/makefiles/rules.make openjdk/hotspot/make/windows/makefiles/rules.make
--- openjdk.orig/hotspot/make/windows/makefiles/rules.make	2013-02-11 23:08:47.000000000 +0000
+++ openjdk/hotspot/make/windows/makefiles/rules.make	2013-02-15 14:53:58.300548666 +0000
@@ -45,7 +45,7 @@
 # Settings for javac
 BOOT_SOURCE_LANGUAGE_VERSION=6
 BOOT_TARGET_CLASS_VERSION=6
-JAVAC_FLAGS=-g -encoding ascii
+JAVAC_FLAGS=-g -encoding ascii -J-Xmx512m
 BOOTSTRAP_JAVAC_FLAGS=$(JAVAC_FLAGS) -source $(BOOT_SOURCE_LANGUAGE_VERSION) -target $(BOOT_TARGET_CLASS_VERSION)
 
 ProjectFile=jvm.vcproj
