--- openjdk.orig/jdk/src/share/classes/java/util/concurrent/atomic/AtomicIntegerArray.java	Tue Nov 17 14:10:36 2009
+++ openjdk/jdk/src/share/classes/java/util/concurrent/atomic/AtomicIntegerArray.java	Tue Nov 17 14:10:35 2009
@@ -57,7 +57,7 @@
     private long rawIndex(int i) {
         if (i < 0 || i >= array.length)
             throw new IndexOutOfBoundsException("index " + i);
-        return base + i * scale;
+        return base + (long) i * scale;
     }
 
     /**
--- openjdk.orig/jdk/src/share/classes/java/util/concurrent/atomic/AtomicLongArray.java	Tue Nov 17 14:10:38 2009
+++ openjdk/jdk/src/share/classes/java/util/concurrent/atomic/AtomicLongArray.java	Tue Nov 17 14:10:37 2009
@@ -56,7 +56,7 @@
     private long rawIndex(int i) {
         if (i < 0 || i >= array.length)
             throw new IndexOutOfBoundsException("index " + i);
-        return base + i * scale;
+        return base + (long) i * scale;
     }
 
     /**
--- openjdk.orig/jdk/src/share/classes/java/util/concurrent/atomic/AtomicReferenceArray.java	Tue Nov 17 14:10:40 2009
+++ openjdk/jdk/src/share/classes/java/util/concurrent/atomic/AtomicReferenceArray.java	Tue Nov 17 14:10:39 2009
@@ -57,7 +57,7 @@
     private long rawIndex(int i) {
         if (i < 0 || i >= array.length)
             throw new IndexOutOfBoundsException("index " + i);
-        return base + i * scale;
+        return base + (long) i * scale;
     }
 
     /**
