diff -r b6701047e287 make/common/shared/Platform.gmk
--- openjdk-ecj.orig/jdk/make/common/shared/Platform.gmk	Sat Jun 12 00:15:32 2010 +0100
+++ openjdk-ecj/jdk/make/common/shared/Platform.gmk	Sun Jun 13 23:54:02 2010 +0100
@@ -88,8 +88,8 @@
 
 SYSTEM_UNAME := $(shell uname)
 
-# Normal boot jdk is previous release, but a hard requirement is a 1.6 boot
-REQUIRED_BOOT_VER = 1.6
+# Normal boot jdk is previous release, but a hard requirement is a 1.5 boot
+REQUIRED_BOOT_VER = 1.5
 
 # If we are using freetype, this is the required version
 REQUIRED_FREETYPE_VERSION=2.3.0
