diff -Nru openjdk-ecj.orig/jdk/make/sun/awt/Makefile openjdk-ecj/jdk/make/sun/awt/Makefile
--- openjdk-ecj.orig/jdk/make/sun/awt/Makefile	2010-08-03 23:52:42.084667732 +0100
+++ openjdk-ecj/jdk/make/sun/awt/Makefile	2010-08-03 23:53:10.308917593 +0100
@@ -34,6 +34,8 @@
 VIS_NEEDED=true
 
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 #
 # Use highest optimization level
diff -Nru openjdk-ecj.orig/jdk/make/sun/xawt/Makefile openjdk-ecj/jdk/make/sun/xawt/Makefile
--- openjdk-ecj.orig/jdk/make/sun/xawt/Makefile	2010-08-03 23:52:56.420668143 +0100
+++ openjdk-ecj/jdk/make/sun/xawt/Makefile	2010-08-03 23:53:18.494167931 +0100
@@ -31,6 +31,8 @@
 PRODUCT = sun
 
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 GEN_DIR=$(GENSRCDIR)/sun/awt/X11
 
diff -Nru openjdk-ecj.orig/jdk/make/javax/sound/Makefile openjdk-ecj/jdk/make/javax/sound/Makefile
--- openjdk-ecj.orig/jdk/make/javax/sound/Makefile	2013-06-24 00:00:22.031254425 +0100
+++ openjdk-ecj/jdk/make/javax/sound/Makefile	2013-06-24 00:01:27.040288124 +0100
@@ -28,6 +28,8 @@
 LIBRARY = jsound
 PRODUCT = sun
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 # include defines for sound
 include SoundDefs.gmk
diff -Nru openjdk-ecj.orig/jdk/make/javax/management/Makefile openjdk-ecj/jdk/make/javax/management/Makefile
--- openjdk-ecj.orig/jdk/make/javax/management/Makefile	2011-11-14 22:11:40.000000000 +0000
+++ openjdk-ecj/jdk/make/javax/management/Makefile	2013-06-24 00:40:59.742001431 +0100
@@ -31,6 +31,8 @@
 PACKAGE = javax.management
 PRODUCT = sun
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 #
 # Files to compile
diff --git a/make/java/awt/Makefile b/make/java/awt/Makefile
--- openjdk-ecj/jdk/make/java/awt/Makefile
+++ openjdk-ecj/jdk/make/java/awt/Makefile
@@ -27,6 +27,8 @@
 PACKAGE = java.awt
 PRODUCT = sun
 include $(BUILDDIR)/common/Defs.gmk
+LANGUAGE_VERSION = -source 6
+CLASS_VERSION = -target 6
 
 
 #
