--- openjdk.orig/jdk/src/share/classes/sun/security/jgss/krb5/InitialToken.java	2010-06-17 12:54:06.000000000 +0800
+++ openjdk/jdk/src/share/classes/sun/security/jgss/krb5/InitialToken.java	2010-06-17 12:54:06.000000000 +0800
@@ -209,6 +209,12 @@
 
             int pos = 0;
 
+            if (checksum == null) {
+                GSSException ge = new GSSException(GSSException.FAILURE, -1,
+                        "No cksum in AP_REQ's authenticator");
+                ge.initCause(new KrbException(Krb5.KRB_AP_ERR_INAPP_CKSUM));
+                throw ge;
+            }
             checksumBytes = checksum.getBytes();
 
             if ((checksumBytes[0] != CHECKSUM_FIRST_BYTES[0]) ||
