--- openjdk.orig/jdk/src/share/classes/sun/java2d/cmm/lcms/LCMSTransform.java	2013-03-07 12:17:44.080685352 +0000
+++ openjdk/jdk/src/share/classes/sun/java2d/cmm/lcms/LCMSTransform.java	2013-03-07 12:39:57.202035054 +0000
@@ -557,18 +557,19 @@
     /* the number of colors is (size of the array) / (number of input/output
        components */
     public short[] colorConvert(short[] src, short[] dst) {
+        LCMSImageLayout srcIL, dstIL;
 
         if (dst == null) {
             dst = new short [(src.length/getNumInComponents())*getNumOutComponents()];
         }
 
         try {
-            LCMSImageLayout srcIL = new LCMSImageLayout(
+            srcIL = new LCMSImageLayout(
                 src, src.length/getNumInComponents(),
                 LCMSImageLayout.CHANNELS_SH(getNumInComponents()) |
                 LCMSImageLayout.BYTES_SH(2), getNumInComponents()*2);
 
-            LCMSImageLayout dstIL = new LCMSImageLayout(
+            dstIL = new LCMSImageLayout(
                 dst, dst.length/getNumOutComponents(),
                 LCMSImageLayout.CHANNELS_SH(getNumOutComponents()) |
                 LCMSImageLayout.BYTES_SH(2), getNumOutComponents()*2);
@@ -584,17 +585,19 @@
     }
 
     public byte[] colorConvert(byte[] src, byte[] dst) {
+        LCMSImageLayout srcIL, dstIL;
+
         if (dst == null) {
             dst = new byte [(src.length/getNumInComponents())*getNumOutComponents()];
         }
 
         try {
-            LCMSImageLayout srcIL = new LCMSImageLayout(
+            srcIL = new LCMSImageLayout(
                 src, src.length/getNumInComponents(),
                 LCMSImageLayout.CHANNELS_SH(getNumInComponents()) |
                 LCMSImageLayout.BYTES_SH(1), getNumInComponents());
 
-            LCMSImageLayout dstIL = new LCMSImageLayout(
+            dstIL = new LCMSImageLayout(
                 dst, dst.length/getNumOutComponents(),
                 LCMSImageLayout.CHANNELS_SH(getNumOutComponents()) |
                 LCMSImageLayout.BYTES_SH(1), getNumOutComponents());
