--- cacao/cacao/src/native/vm/sun_misc_Unsafe.c.orig	2010-01-11 16:20:51.000000000 +0000
+++ cacao/cacao/src/native/vm/sun_misc_Unsafe.c	2010-01-11 16:22:15.000000000 +0000
@@ -107,9 +107,10 @@
 #if 0
 	/* OpenJDK 7 */
 	{ "setMemory",              "(Ljava/lang/Object;JJB)V",                                   (void *) (intptr_t) &Java_sun_misc_Unsafe_setMemory                      },
+#endif
 	{ "copyMemory",             "(Ljava/lang/Object;JLjava/lang/Object;JJ)V",                 (void *) (intptr_t) &Java_sun_misc_Unsafe_copyMemory                     },
-#else
 	{ "setMemory",              "(JJB)V",                                                     (void *) (intptr_t) &Java_sun_misc_Unsafe_setMemory                      },
+#if 0
 	{ "copyMemory",             "(JJJ)V",                                                     (void *) (intptr_t) &Java_sun_misc_Unsafe_copyMemory                     },
 #endif
 	{ "freeMemory",             "(J)V",                                                       (void *) (intptr_t) &Java_sun_misc_Unsafe_freeMemory                     },
@@ -786,7 +787,7 @@
 
 	system_memset(p, value, length);
 }
-
+#endif
 
 /*
  * Class:     sun/misc/Unsafe
@@ -816,7 +817,7 @@
 
 	system_memcpy(dest, src, length);
 }
-#else
+
 /*
  * Class:     sun/misc/Unsafe
  * Method:    setMemory
@@ -841,7 +842,7 @@
 	system_memset(p, value, length);
 }
 
-
+#if 0
 /*
  * Class:     sun/misc/Unsafe
  * Method:    copyMemory
