diff -ur openjdk.orig/hotspot/src/share/vm/opto/loopnode.cpp openjdk/hotspot/src/share/vm/opto/loopnode.cpp
--- openjdk.orig/hotspot/src/share/vm/opto/loopnode.cpp	2011-11-14 17:07:36.000000000 -0500
+++ openjdk/hotspot/src/share/vm/opto/loopnode.cpp	2012-01-30 11:25:10.648080484 -0500
@@ -546,7 +546,7 @@
 Node* CountedLoopNode::match_incr_with_optional_truncation(
                       Node* expr, Node** trunc1, Node** trunc2, const TypeInt** trunc_type) {
   // Quick cutouts:
-  if (expr == NULL || expr->req() != 3)  return false;
+  if (expr == NULL || expr->req() != 3)  return NULL;
 
   Node *t1 = NULL;
   Node *t2 = NULL;
