diff -Nru openjdk.orig/jdk/make/common/Defs-linux.gmk openjdk/jdk/make/common/Defs-linux.gmk
--- openjdk.orig/jdk/make/common/Defs-linux.gmk	2015-01-21 21:25:32.000282631 +0000
+++ openjdk/jdk/make/common/Defs-linux.gmk	2015-01-21 21:27:09.061441927 +0000
@@ -115,7 +115,7 @@
 CFLAGS_REQUIRED_m68k    +=
 CFLAGS_REQUIRED_mips    +=
 CFLAGS_REQUIRED_mipsel  += -D_LITTLE_ENDIAN
-CFLAGS_REQUIRED_ppc     += -m32
+CFLAGS_REQUIRED_ppc     += -m32 -fsigned-char -D_BIG_ENDIAN
 CFLAGS_REQUIRED_ppc64   += -m64
 CFLAGS_REQUIRED_s390    +=
 CFLAGS_REQUIRED_s390x   += -m64
@@ -160,11 +160,7 @@
 #
 # Misc compiler options
 #
-ifeq ($(ARCH),ppc)
-  CFLAGS_COMMON   = -fsigned-char
-else # ARCH
-  CFLAGS_COMMON   = -fno-strict-aliasing
-endif # ARCH
+CFLAGS_COMMON  = -fno-strict-aliasing
 PIC_CODE_LARGE = -fPIC
 PIC_CODE_SMALL = -fpic
 GLOBAL_KPIC = $(PIC_CODE_LARGE)
