diff --git a/make/linux/makefiles/rules.make b/make/linux/makefiles/rules.make
--- openjdk/hotspot/make/linux/makefiles/rules.make
+++ openjdk/hotspot/make/linux/makefiles/rules.make
@@ -126,8 +126,8 @@
 RUN.JAR$(MAKE_VERBOSE) += >/dev/null
 
 # Settings for javac
-BOOT_SOURCE_LANGUAGE_VERSION = 5
-BOOT_TARGET_CLASS_VERSION = 5
+BOOT_SOURCE_LANGUAGE_VERSION = 6
+BOOT_TARGET_CLASS_VERSION = 6
 JAVAC_FLAGS = -g -encoding ascii
 BOOTSTRAP_JAVAC_FLAGS = $(JAVAC_FLAGS) -source $(BOOT_SOURCE_LANGUAGE_VERSION) -target $(BOOT_TARGET_CLASS_VERSION)
 
diff --git a/make/solaris/makefiles/rules.make b/make/solaris/makefiles/rules.make
--- openjdk/hotspot/make/solaris/makefiles/rules.make
+++ openjdk/hotspot/make/solaris/makefiles/rules.make
@@ -118,8 +118,8 @@
 RUN.JAR$(MAKE_VERBOSE) += >/dev/null
 
 # Settings for javac
-BOOT_SOURCE_LANGUAGE_VERSION = 5
-BOOT_TARGET_CLASS_VERSION = 5
+BOOT_SOURCE_LANGUAGE_VERSION = 6
+BOOT_TARGET_CLASS_VERSION = 6
 JAVAC_FLAGS = -g -encoding ascii
 BOOTSTRAP_JAVAC_FLAGS = $(JAVAC_FLAGS) -source $(BOOT_SOURCE_LANGUAGE_VERSION) -target $(BOOT_TARGET_CLASS_VERSION)
 
diff --git a/make/windows/makefiles/rules.make b/make/windows/makefiles/rules.make
--- openjdk/hotspot/make/windows/makefiles/rules.make
+++ openjdk/hotspot/make/windows/makefiles/rules.make
@@ -43,8 +43,8 @@
 !endif
 
 # Settings for javac
-BOOT_SOURCE_LANGUAGE_VERSION=5
-BOOT_TARGET_CLASS_VERSION=5
+BOOT_SOURCE_LANGUAGE_VERSION=6
+BOOT_TARGET_CLASS_VERSION=6
 JAVAC_FLAGS=-g -encoding ascii
 BOOTSTRAP_JAVAC_FLAGS=$(JAVAC_FLAGS) -source $(BOOT_SOURCE_LANGUAGE_VERSION) -target $(BOOT_TARGET_CLASS_VERSION)
 
