diff --git a/agent/src/share/classes/sun/jvm/hotspot/gc_implementation/g1/HeapRegionSeq.java b/agent/src/share/classes/sun/jvm/hotspot/gc_implementation/g1/HeapRegionSeq.java
--- openjdk/hotspot/agent/src/share/classes/sun/jvm/hotspot/gc_implementation/g1/HeapRegionSeq.java
+++ openjdk/hotspot/agent/src/share/classes/sun/jvm/hotspot/gc_implementation/g1/HeapRegionSeq.java
@@ -77,13 +77,10 @@
         private long index;
         private long length;
 
-        @Override
         public boolean hasNext() { return index < length; }
 
-        @Override
         public HeapRegion next() { return at(index++);    }
 
-        @Override
         public void remove()     { /* not supported */    }
 
         HeapRegionIterator(Address addr) {
