diff -Nru openjdk.orig/hotspot/src/share/vm/runtime/os.cpp openjdk/hotspot/src/share/vm/runtime/os.cpp
--- openjdk.orig/hotspot/src/share/vm/runtime/os.cpp	2013-06-04 18:47:35.000000000 +0100
+++ openjdk/hotspot/src/share/vm/runtime/os.cpp	2013-08-15 11:53:40.989493660 +0100
@@ -1105,6 +1105,9 @@
 #ifdef __APPLE__
         "%/lib/JObjC.jar:"
 #endif
+        "%/lib/netx.jar:"
+        "%/lib/plugin.jar:"
+        "%/lib/rhino.jar:"
         "%/classes";
     char* sysclasspath = format_boot_path(classpath_format, home, home_len, fileSep, pathSep);
     if (sysclasspath == NULL) return false;
